Add `jnp.resize`

lgeiger opened the pull request:
This PR adds `jnp.resize` to match [`np.resize`](https://numpy.org/doc/stable/reference/generated/numpy.resize.html).

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -1415,6 +1415,24 @@ def unravel_index(indices, shape):
+@_wraps(np.resize)
+def resize(a, new_shape):
+  if isinstance(new_shape, (int, integer)):
+    new_shape = (new_shape,)
```
We have the `_ensure_index_tuple` utility for this check, which is a bit more robust.

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -1415,6 +1415,24 @@ def unravel_index(indices, shape):
+    return zeros_like(a, shape=new_shape)
+
+  repeats = ceil_of_ratio(new_size, a.size)
+  a = concatenate((a,) * repeats)[:new_size]
```
I suspect `tile(a, repeats)[:new_size]` will be more efficient, because it's based on broadcasting rather than concatenation.

jakevdp commented on tests/lax_numpy_test.py:
```diff
@@ -3166,6 +3166,31 @@ def testReshapeMethod(self, arg_shape, out_shape, dtype):
+          ((2, 2), (0,)),
+          ((2, 2), (0, 2)),
+          ((2, 2), (2, 0)),
+      ]))
```
It would be good to also test empty arrays and scalar shapes. How about `for arg_shape, out_shape in zip(all_shapes, all_shapes)`?

jakevdp replied:
or if you want to be fancy, `itertools.combinations_with_replaecement(all_shapes, 2)`

jakevdp replied:
Or maybe it's `itertools.product(all_shapes, 2)`. Thinking is hard

lgeiger replied:
`zip(all_shapes, all_shapes)` would not work here, since we want to test the actual resizing with different shapes. However I can change it to `itertools.product(all_shapes, repeat=2)` if it doesn't matter that it generates quite a lot of test cases. Alternatively I can also remove the parameterization on `dtype` which would reduce the amount of generated tests which is probably fine since there is no `dtype` specific code in `jnp.resize`. Let me know what you prefer.

lgeiger replied:
Sorry, I missed your previous comments. I changed it too `itertools.product([0] + array_shapes, repeat=2)`. Unfortunately we can't use  `all_shapes` here, since the numpy reference implementation can't handle `jtu.NUMPY_SCALAR_SHAPE` and  `jtu.PYTHON_SCALAR_SHAPE`.

jakevdp replied:
Right, that makes sense.

I think we could probably use `all_shapes` on the left, sinc they are never passed to numpy. so something like: `for arg_shape, out_shape in itertools.product(all_shapes, array_shapes)`. What do you think?

lgeiger replied:
Good idea, I changed it.

jakevdp reviewed (changes requested):
Looks great, thanks! Just a couple minor comments below...

jakevdp reviewed (approved):
Very nice, thanks!

hawkinsp commented:
This PR appears to cause test failures for Python 3.6 and NumPy 1.17.5:

```
$ pytest tests/lax_numpy_test.py -k testResize --tb=short
========================================= test session starts =========================================
platform darwin -- Python 3.6.12, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /Users/phawkins/p/jaxlg, configfile: pytest.ini
plugins: xdist-2.1.0, forked-1.3.0
collected 4145 items / 4135 deselected / 10 selected

tests/lax_numpy_test.py .......F.F                                                              [100%]

============================================== FAILURES ===============================================
_______________ LaxBackedNumpyTests.testResize_inshape=float32[3,1]_outshape=float32[] ________________
tests/lax_numpy_test.py:3181: in testResize
    self._CheckAgainstNumpy(np_fun, jnp_fun, args_maker)
jax/test_util.py:991: in _CheckAgainstNumpy
    numpy_ans = numpy_reference_op(*args)
tests/lax_numpy_test.py:3178: in <lambda>
    np_fun = lambda x: np.resize(x, out_shape)
<__array_function__ internals>:6: in resize
    ???
../../.pyenv/versions/3.6.12/envs/py36/lib/python3.6/site-packages/numpy/core/fromnumeric.py:1372: in resize
    a = a[:-extra]
E   TypeError: slice indices must be integers or None or have an __index__ method
___________________ LaxBackedNumpyTests.testResize_inshape=int8[4]_outshape=int8[] ____________________
tests/lax_numpy_test.py:3181: in testResize
    self._CheckAgainstNumpy(np_fun, jnp_fun, args_maker)
jax/test_util.py:991: in _CheckAgainstNumpy
    numpy_ans = numpy_reference_op(*args)
tests/lax_numpy_test.py:3178: in <lambda>
    np_fun = lambda x: np.resize(x, out_shape)
<__array_function__ internals>:6: in resize
    ???
../../.pyenv/versions/3.6.12/envs/py36/lib/python3.6/site-packages/numpy/core/fromnumeric.py:1372: in resize
    a = a[:-extra]
E   TypeError: slice indices must be integers or None or have an __index__ method
======================================= short test summary info =======================================
FAILED tests/lax_numpy_test.py::LaxBackedNumpyTests::testResize_inshape=float32[3,1]_outshape=float32[]
FAILED tests/lax_numpy_test.py::LaxBackedNumpyTests::testResize_inshape=int8[4]_outshape=int8[] - Ty...
============================ 2 failed, 8 passed, 4135 deselected in 3.40s =============================
```

I'm not sure why this failure doesn't show up in CI.

lgeiger committed:
Add `jnp.resize`

lgeiger commented:
> This PR appears to cause test failures for Python 3.6 and NumPy 1.17.5:

@hawkinsp Good catch, it looks like empty output shapes are only supported since https://github.com/numpy/numpy/commit/7ef6b65fa960fcedcef8a27b1c3ef8c67ebb2078 in Numpy 1.20. I fixed the unittests and rebased the PR.

jakevdp commented:
Looks like it worked - thanks!
